src/NLog/Targets/TargetWithContext.cs

```diff
@@ -0,0 +1,797 @@
+            OptimizeBufferReuse = true;
+        }
+
+        /// <summary>
+        /// Check if logevent has properties (or context properties)
+        /// </summary>
+        /// <param name="logEvent"></param>
+        /// <returns>True if properties should be included</returns>
+        protected bool HasCombinedProperties(LogEventInfo logEvent)
+        {
+            return IncludeGdc
+            || IncludeMdc
+#if !SILVERLIGHT
+            || IncludeMdlc
+#endif
+            || (IncludeEventProperties && (logEvent?.HasProperties ?? false));
+        }
+
+        /// <summary>
+        /// Checks if any context properties, and if any returns them as a single dictionary
+        /// </summary>
+        /// <param name="logEvent"></param>
+        /// <returns>Dictionary with any context properties for the logEvent (Null if none found)</returns>
+        protected IDictionary<string, object> LookupContextProperties(LogEventInfo logEvent)
+        {
+            IDictionary<string, object> combinedPropties = null;
+            if (IncludeGdc)
+            {
+                combinedPropties = CaptureContextGdc(logEvent, combinedPropties);
+            }
+
+            if (IncludeMdc)
+            {
+                if (!CombineProperties(logEvent, _contextLayout.MdcLayout, ref combinedPropties))
+                {
+                    combinedPropties = CaptureContextMdc(logEvent, combinedPropties);
+                }
+            }
+
+#if !SILVERLIGHT
+            if (IncludeMdlc)
+            {
+                if (!CombineProperties(logEvent, _contextLayout.MdlcLayout, ref combinedPropties))
+                {
+                    combinedPropties = CaptureContextMdlc(logEvent, combinedPropties);
+                }
+            }
+#endif
+
+            if (ContextProperties != null && ContextProperties.Count > 0)
+            {
+                combinedPropties = combinedPropties ?? new Dictionary<string, object>();
+                for (int i = 0; i < ContextProperties.Count; ++i)
+                {
+                    var attrib = ContextProperties[i];
+                    if (attrib == null || string.IsNullOrEmpty(attrib.Name) || attrib.Layout == null)
+                        continue;
+
+                    var attribValue = RenderLogEvent(attrib.Layout, logEvent);
+                    if (attrib.ExcludeEmptyValue && string.IsNullOrEmpty(attribValue))
+                        continue;
+
+                    combinedPropties[attrib.Name] = attribValue;
+                }
+            }
+
+            return combinedPropties;
+        }
+
+        /// <summary>
+        /// Creates combined dictionary of all configured properties for logEvent
+        /// </summary>
+        /// <param name="logEvent"></param>
+        /// <returns>Dictionary with all collected properties for logEvent</returns>
+        protected IDictionary<string, object> CreateCombinedProperties(LogEventInfo logEvent)
```

Comment:
proposal rename to GetCombinedProperties or maybe even GetAllProperties.

Reply:
Done